### PR TITLE
Add a DynamicReader interface to serve generated strings

### PR DESCRIPTION
**Pull request by mont5piques**
Hi,

I added a piece of code that allows to serve dynamic generated strings in addition of static files located in tftp root. I added a new class inpired from FilesystemReader to do so.

I hope you will accept my PR, beacause I think your tftp implementation is one of the best for Python, especially for the twisted backend.

I'm using it in Nytrio for generating dynamic grub bootmenus (take a look at [http://nytrio.com](http://nytrio.com) if you have some time).

Thanks,

### CODE REVIEW

**Comment by shylent** [comment]
Hi! Thanks for the PR! Can I persuade you to write a test case with unit tests for new functionality?

**Comment by mont5piques** [comment]
Okay, I'll add a test case in test_backend.py

**Comment by shylent** [comment]
I haven't forgotten about this PR, by the way, I am just waiting for the tests.

File: tftp/backend.py [hunk 1]
```diff
@@ -231,6 +231,72 @@ def cancel(self):
...
+@interface.implementer(IReader)
+class DynamicReader(object):
+    """A reader to go with L{FilesystemSynchronousBackend}.
+
+    @see: L{IReader}
+
+    @param file_path: relative file path, specified as part of a TFTP read request (RRQ)
+    @type file_path: C{bytes}
+
+    @param dyn_handler: a callback function to handle dynamic requests
```
Comment (shylent) [inline tftp/backend.py]: The documentation must include info w.r.t. what is expected of this callback function (what arguments it must accept, what does it have to do etc).

File: tftp/backend.py [hunk 2]
```diff
@@ -270,7 +336,16 @@ def get_reader(self, file_name):
...
-        return FilesystemReader(target_path)
+        try:
+            # Try using FilesystemReader first
+            return FilesystemReader(target_path)
+        except FileNotFound:
+            # Try using DynamicReader if dyn_handler was supplied
+            if self.dyn_handler is not None:
+                return DynamicReader(file_name, self.dyn_handler)
+            else:
+                raise
```
Comment (shylent) [inline tftp/backend.py]: Out of curiosity: do you really need to have a backend, that can both write to filesystem and "read" using a DynamicReader? Maybe you need to create a separate `IBackend` implementation, that always returns a `DynamicReader` in `get_reader` and leave `FilesystemSynchronousBackend` as it is now?

**Pull request closed without merging**